From 4d3e54f95d8a5c1aa4457d3707316637b8fdd80e Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Wed, 4 Jun 2014 20:06:43 -0400
Subject: [net] tcp: syncookies: do not use getnstimeofday()

Message-id: <1401912404-10861-4-git-send-email-fwestpha@redhat.com>
Patchwork-id: 82577
O-Subject: [PATCH 3/4] [RHEL7.1 net] tcp: syncookies: do not use getnstimeofday()
Bugzilla: 1084439
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hsowa@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>

cherry-pick.

commit 632623153196bf183a69686ed9c07eee98ff1bf8
Author: Eric Dumazet <edumazet@google.com>
Date:   Wed Mar 19 21:02:21 2014 -0700

    tcp: syncookies: do not use getnstimeofday()

    While it is true that getnstimeofday() uses about 40 cycles if TSC
    is available, it can use 1600 cycles if hpet is the clocksource.

    Switch to get_jiffies_64(), as this is more than enough, and
    go back to 60 seconds periods.

    Fixes: 8c27bd75f04f ("tcp: syncookies: reduce cookie lifetime to 128 seconds")
    Signed-off-by: Eric Dumazet <edumazet@google.com>
    Cc: Florian Westphal <fw@strlen.de>
    Acked-by: Florian Westphal <fw@strlen.de>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/net/tcp.h b/include/net/tcp.h
index b0b678f..8ab2902 100644
--- a/include/net/tcp.h
+++ b/include/net/tcp.h
@@ -489,20 +489,21 @@ extern struct sock *cookie_v4_check(struct sock *sk, struct sk_buff *skb,
         struct ip_options *opt);
 #ifdef CONFIG_SYN_COOKIES
 #include <linux/ktime.h>
-/* Syncookies use a monotonic timer which increments every 64 seconds.
+/* Syncookies use a monotonic timer which increments every 60 seconds.
  * This counter is used both as a hash input and partially encoded into
  * the cookie value.  A cookie is only validated further if the delta
  * between the current counter value and the encoded one is less than this,
- * i.e. a sent cookie is valid only at most for 128 seconds (or less if
+ * i.e. a sent cookie is valid only at most for 2*60 seconds (or less if
  * the counter advances immediately after a cookie is generated).
  */
 #define MAX_SYNCOOKIE_AGE 2
 
 static inline u32 tcp_cookie_time(void)
 {
- struct timespec now;
- getnstimeofday(&now);
- return now.tv_sec >> 6; /* 64 seconds granularity */
+ u64 val = get_jiffies_64();
+
+ do_div(val, 60 * HZ);
+ return val;
 }
 
 extern u32 __cookie_v4_init_sequence(const struct iphdr *iph,
-- 
1.7.1